### ISSUE TITLE
Application logs not sending from Databricks cluster to Application Insight.

### ISSUE THREAD

**Issue by jeeveshpandey14** [microsoft/ApplicationInsights-Java#802]
### Expected behavior
I am using applicationinsights-logging-log4j1_2 for sending all logs to application insight. When I uploaded my jar as a library in databricks cluster,  it is not sending logs to application insight.

#### Sample Application

Dependencies:
```sh
    compile 'log4j:log4j:1.2.17'
    compile 'commons-logging:commons-logging:1.2'
    compile group: 'com.microsoft.azure', name: 'applicationinsights-web', version: '2.+'
    compile group: 'com.microsoft.azure', name: 'applicationinsights-logging-log4j1_2', version: '2.0.+'
```

This is my log4j configuration:
```sh
<!DOCTYPE log4j:configuration SYSTEM "log4j.dtd">
<log4j:configuration debug="true" xmlns:log4j='http://jakarta.apache.org/log4j/'>

    <appender name="console" class="org.apache.log4j.ConsoleAppender">
        <param name="Target" value="System.out"/>
        <layout class="org.apache.log4j.PatternLayout">
            <param name="ConversionPattern" value="%d{yyyy-MM-dd HH:mm:ss} %-5p %c{1}:%L - %m%n" />
        </layout>
    </appender>
    <appender name="aiAppender"
              class="com.microsoft.applicationinsights.log4j.v1_2.ApplicationInsightsAppender">
    </appender>

    <root>
        <priority value ="trace"/>
        <appender-ref ref="console"/>
        <appender-ref ref="aiAppender" />
    </root>

</log4j:configuration>
```

### System information
Please provide the following information:
 - SDK Version: applicationinsights-logging-log4j1_2
 - OS type and version: Databricks cluster
 - Application Server type and version (if applicable):
 - Using spring-boot? - Simple Java Project

**Comment by dhaval24** [comment]
@jeeveshpandey14 can you please explicitly specify the latest SDK 2.3.0 and also turn on SDKLogs and let us know.

**Comment by dhaval24** [comment]
@jeeveshpandey14 can you please provide SDK Logs. You can turn it on using the https://github.com/Microsoft/ApplicationInsights-Java/wiki/ApplicationInsights.XML#sdklogger and upload it here. I would like to know if there are any errors coming from SDK.

**Comment by nuclear0wl** [comment]
Hello all, I was faced with the same issue. @dhaval24 and @jeeveshpandey14 do you have any updated?

**Comment by dhaval24** [comment]
@nuclear0wl thanks for asking. Would it be possible to provide a small repro to investigate this further?

**Comment by nuclear0wl** [comment]
We are trying to run our spark streaming application and send driver/executors logs into Application Insights service.
Our architecture consists of several modules:
1. main "application.jar" file aka spark streaming application
2. "application-insights-client.jar" file, that includes applicationinsights-logging-log4j1_2 dependency, MyCustomContextInitializer class and ApplicationInsights.xml file. MyCustomContextInitializer class provides possibility to add custom properties to logs (property.one, property.two... see below).
MyCustomContextInitializer:
```
public class AppInsightsConfigurableContextInitializer implements ContextInitializer {
    private Map<String, String> props;
    public AppInsightsConfigurableContextInitializer() {
        this.props = new HashMap<>();
        props.put("PropertyOne", System.getProperty("property.one"));
        props.put("PropertyTwo", System.getProperty("property.one"));
    }
    public void initialize(TelemetryContext context) {
        context.getProperties().putAll(props);
    }
}
```

- ApplicationInsights.xml:
```
<?xml version="1.0" encoding="utf-8"?>
<ApplicationInsights xmlns="http://schemas.microsoft.com/ApplicationInsights/2013/Settings" schemaVersion="2014-05-30">
    <!-- The key from the portal: -->
    <InstrumentationKey>***key***</InstrumentationKey>
    <ContextInitializers>
        <Add type="path.to.class.MyCustomContextInitializer"/>
    </ContextInitializers>
</ApplicationInsights>
3. "log4j.properties" file, that includes several appenders: ConsoleAppender, RollingFileAppender and appender for Application Insights as well.
- log4j.properties
log4j.appender.aiAppender=com.microsoft.applicationinsights.log4j.v1_2.ApplicationInsightsAppender
log4j.appender.aiAppender.layout=org.apache.log4j.PatternLayout
```

Our spark-submit script looks like this:
```
[  
   "--class",
   "path.to.main.class.Driver",
   "--name",
   "MyName",
   "--conf",
   "spark.logConf=true",
   "--conf",
   "spark.serializer=org.apache.spark.serializer.KryoSerializer",
   "--conf",
   "spark.driver.extraJavaOptions=-Dlog4j.configuration=file:'/dbfs/mnt/my_adl/distr/log4j.properties' -Dproperty.one=DriverOne -Dproperty.two=DriverTwo",
   "--conf",
   "spark.executor.extraJavaOptions=-Dlog4j.configuration=file:'/dbfs/mnt/my_adl/distr/log4j.properties' -Dproperty.one=ExecutorOne -Dproperty.two=ExecutorTwo",
   "dbfs:/mnt/my_adl/distr/application.jar",
   "--somePropName",
   "propValue"
]
```

If we use -DAPPLICATION_INSIGHTS_IKEY option, we will find logs in Application Insights service, however we wont see any custom properties from custom context initializer.
If we don't use -DAPPLICATION_INSIGHTS_IKEY option, we wont see anything in Application Insights service. So we deside that ApplicationInsights.xml file ignored by logger.

How should we use it properly?

**Comment by dhaval24** [comment]
@nuclear0wl it looks like you are not packaging `applicationinsights-core` jar. Can you please confirm whether you have core as well as logging dependency in the project?

**Comment by nuclear0wl** [comment]
@dhaval24 we use jar `applicationinsights-logging-log4j1_2` as dependency, it's already contains `applicationinsights-core`.

**Comment by dhaval24** [comment]
@nuclear0wl yes, it looks like ApplicationInsights.xml is being ignored. It might be due to how spark application bootstraps. Please turn on the SDKLogger by adding the tag:
`<SDKLogger />` in the ApplicationInsights.xml and see if you get any console logs. If you are not getting any console logs, that means probably yes the config file is getting ignored and we might need more debugging in this case.

**Comment by nuclear0wl** [comment]
@dhaval24 we've already done that and didn't receive any logs. Also, this solution works properly on Azure HDInsight and we don't mind why it isn't fits Azure Databricks. 
Also, we've set property `Dapplicationinsights.configurationDirectory` and that also don't work.

**Comment by nuclear0wl** [comment]
Hello all, we've found root cause of considering issue. Particularly, we don't have pre-delivered jaxb on our Databricks cluster, however we have listed below dependencies on HDInsight cluster out of box.
 
At first, we can read ApplicationInsights.xml file, but without jaxb library we can't parse it.
Moreover, we can't turn on SDKLogger without jaxb.
 
So, adding of this dependencies solves our issue:
``` 
<dependency>
                <groupId>com.sun.xml.bind</groupId>
                <artifactId>jaxb-impl</artifactId>
                <version>2.3.2</version>
</dependency>
<dependency>
                <groupId>com.sun.xml.bind</groupId>
                <artifactId>jaxb-core</artifactId>
                <version>2.3.0.1</version>
</dependency>
```

**Comment by DJLemkes** [comment]
Ran into the same kind of issue when using this library in a Scala project. Everything worked fine until we started packaging our application into a Docker container running OpenJDK 11.

As of Java 11, the [JAXB dependencies should be included manually](https://www.jesperdj.com/2018/09/30/jaxb-on-java-9-10-11-and-beyond/). Because we let SBT compile with Java 8 (as is recommended), this resulted in runtime `NoClassDefFound` exceptions. 

At first, just including `jaxb-api` seemed to solve the problem because it removed the error. But, then it just started to silently fail and fall back to default config. Turning on `SKDLogger` didn't make any difference because the config XML was not being parsed.

Including the following dependencies solved our problem:

```scala
"javax.xml.bind"    % "jaxb-api"             % "2.3.1",
"com.sun.xml.bind"  % "jaxb-core"            % "2.3.0.1",
"com.sun.xml.bind"  % "jaxb-impl"            % "2.3.2",
```
